### PR TITLE
Fixed #24423 -- Reorganized i18n tag tests.

### PR DESCRIPTION
**Pull request by accraze**
Piggy-backing off of work done in the closed PR #7363:

* Moved all tag tests into syntax_tests/i18n:
```
├── __init__.py
├── base.py
├── test_blocktrans.py
├── test_get_available_languages.py
├── test_get_language_info.py
├── test_get_language_info_list.py
├── test_trans.py
└── test_underscore_syntax.py
```

### CODE REVIEW

File: tests/template_tests/syntax_tests/i18n/test_get_language_info_list.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# coding: utf-8
```
Comment (timgraham) [inline tests/template_tests/syntax_tests/i18n/test_get_language_info_list.py]: Please use the same ` -*- coding: utf-8 -*-` in all files.

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: tests/i18n/tests.py [hunk 2]
```diff
@@ -68,8 +68,10 @@ def test_plural(self):
         """
         self.assertEqual(ungettext("%d year", "%d years", 0) % 0, "0 année")
         self.assertEqual(ungettext("%d year", "%d years", 2) % 2, "2 années")
-        self.assertEqual(ungettext("%(size)d byte", "%(size)d bytes", 0) % {'size': 0}, "0 octet")
-        self.assertEqual(ungettext("%(size)d byte", "%(size)d bytes", 2) % {'size': 2}, "2 octets")
+        self.assertEqual(ungettext("%(size)d byte", "%(size)d bytes", 0) % {
```
Comment (timgraham) [inline tests/i18n/tests.py]: Could you please revert the unrelated line length changes? I pushed some fixes to your branch before I realized that this file is also changed. Maybe your IDE went wild? We allow longer lines per https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/.

File: tests/i18n/tests.py [hunk 3]
```diff
@@ -1658,132 +1578,6 @@ def test_fallback_language_code(self):
         self.assertEqual(li['code'], 'zh-hans')
 
 
-class MultipleLocaleActivationTests(SimpleTestCase):
```
Comment (timgraham) [inline tests/i18n/tests.py]: I'm not sure if duplicating this class 3 times is the best idea. At least, a common base class could be created. Maybe it's easiest to put trans/blocktrans/_ together. Or all "_" tests could go in a file rather than split between test_multiple_locales.py and test_string_literals.py.

File: tests/template_tests/syntax_tests/i18n/test_blocktrans.py [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+
+    @setup({'i18n27': '{% load i18n %}{% blocktrans count counter=number %}'
+                      '{{ counter }} result{% plural %}{{ counter }} results'
+                      '{% endblocktrans %}'})
+    def test_i18n27(self):
+        """translation of singular form in Russian (#14126)"""
+        with translation.override('ru'):
+            output = self.engine.render_to_string('i18n27', {'number': 1})
+        self.assertEqual(
+            output, '1 \u0440\u0435\u0437\u0443\u043b\u044c\u0442\u0430\u0442')
```
Comment (timgraham) [inline tests/template_tests/syntax_tests/i18n/test_blocktrans.py]: Wrapping wasn't reverted here (I didn't check the patch exhaustively perhaps you can give it another look).

File: tests/template_tests/syntax_tests/i18n/test_trans.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    @setup({'i18n24': '{% load i18n %}{% trans \'Page not found\'|upper %}'})
+    def test_i18n24(self):
+        with translation.override('de'):
+            output = self.engine.render_to_string('i18n24')
+        self.assertEqual(output, 'SEITE NICHT GEFUNDEN')
+
+    @setup({'i18n25': '{% load i18n %}{% trans somevar|upper %}'})
+    def test_i18n25(self):
+        with translation.override('de'):
+            output = self.engine.render_to_string(
```
Comment (timgraham) [inline tests/template_tests/syntax_tests/i18n/test_trans.py]: fix

File: tests/template_tests/syntax_tests/i18n/test_trans.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+
+    @setup({'i18n20': '{% load i18n %}{% trans andrew %}'})
+    def test_i18n20(self):
+        output = self.engine.render_to_string('i18n20', {'andrew': 'a & b'})
+        self.assertEqual(output, 'a &amp; b')
+
+    @setup({'i18n22': '{% load i18n %}{% trans andrew %}'})
+    def test_i18n22(self):
+        output = self.engine.render_to_string(
+            'i18n22', {'andrew': mark_safe('a & b')})
```
Comment (timgraham) [inline tests/template_tests/syntax_tests/i18n/test_trans.py]: fix

File: tests/template_tests/syntax_tests/i18n/test_trans.py [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+from django.template import Context, Template, TemplateSyntaxError
+from django.test import SimpleTestCase, override_settings
+from django.utils import translation
+from django.utils._os import upath
+from django.utils.safestring import mark_safe
+from django.utils.translation import activate, get_language, trans_real
+
+from ...utils import setup
+
+here = os.path.dirname(os.path.dirname(os.path.abspath(upath(__file__))))
```
Comment (timgraham) [inline tests/template_tests/syntax_tests/i18n/test_trans.py]: It would be better to factor out shared elements such as these variables and MultipleLocaleActivationTests setUp/tearDown into a `base.py`.

**Comment by accraze** [comment]
@timgraham I made a `base.py` and extracted the duplicated variables and setUp/tearDown functions. Let me know if anything else needs to be fixed.

File: tests/template_tests/syntax_tests/i18n/base.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+from django.utils.translation import activate, get_language
+
+here = os.path.dirname(os.path.dirname(os.path.abspath(upath(__file__))))
+pdir = os.path.split(os.path.split(os.path.abspath(here))[0])[0]
+extended_locale_paths = settings.LOCALE_PATHS + [
+    os.path.join(pdir, 'i18n', 'other', 'locale'),
+]
+
+
+class BaseTestCase(SimpleTestCase):
```
Comment (timmartin) [inline tests/template_tests/syntax_tests/i18n/base.py]: I feel like there could be a better class name here, to indicate that this isn't a generic base test case. Perhaps `I18nTestCase` would do? I realise it's not a big deal since no other tests should import this `base.py`, but it's useful documentation to someone reading the code. There's also precedent for this in e.g. `WidgetTest` and `SiteMapTestsBase`.

**Comment by accraze** [comment]
Thanks for the feedback @timmartin! I agree that `I18nTestCase` is a better name.
